torch/onnx/_internal/onnxruntime.py

```diff
@@ -828,56 +912,100 @@ def compile(self, graph_module: torch.fx.GraphModule, args) -> torch.fx.GraphMod
+    ) -> "OrtBackend":
+        """Returns a possibly cached instance of an ``OrtBackend``. If an existing
+        backend was created previously through this function with the same options,
+        it will be returned. Otherwise a new backend will be created, cached, and
+        returned.
+
+        Note: if ``options`` sets ``ort_session_options``, a new ``OrtBackend``
+        will always be returned, since ``onnxruntime.SessionOptions`` cannot
+        participate in caching."""
+
+        def reusable(a: OrtBackendOptions, b: OrtBackendOptions):
+            if (
+                a.preferred_execution_providers != b.preferred_execution_providers
+                or a.infer_execution_providers != b.infer_execution_providers
+                or a.default_execution_providers != b.default_execution_providers
+                or a.preallocate_output != b.preallocate_output
+                or a.use_aot_autograd != b.use_aot_autograd
+            ):
+                return False
+
+            # onnxruntime.SessionOptions is a pybind11 object, cannot be pickled,
+            # and holds too much potential state to reasonably check manually;
+            # ort_session_options is provided at all, the backend does not participate
+            # in caching.
+            if a.ort_session_options is not None or b.ort_session_options is not None:
+                return False
+
+            if a.export_options is b.export_options:
+                return True
+
+            # Similarly, some objects in ExportOptions are too stateful to use for
+            # caching. We should revisit this.
+            if a.export_options is not None and b.export_options is not None:
+                return (
+                    a.export_options.dynamic_shapes == b.export_options.dynamic_shapes
+                    and a.export_options.op_level_debug
+                    == b.export_options.op_level_debug
+                    and a.export_options.diagnostic_options
+                    == b.export_options.diagnostic_options
+                    and a.export_options.onnx_registry is b.export_options.onnx_registry
+                    and a.export_options.fake_context is b.export_options.fake_context
+                )
+
+            return False
+
+        if not isinstance(options, OrtBackendOptions):
+            options = OrtBackendOptions(**(options or {}))
+
+        backend = next(
+            (b for b in OrtBackend.__instance_cache if reusable(b._options, options)),
+            None,
+        )
+
+        if backend is None:
+            OrtBackend.__instance_cache.append(backend := OrtBackend(options))
+
+        return backend
+
+    @staticmethod
+    def clear_cached_instances():
+        OrtBackend.__instance_cache.clear()
+
+    @staticmethod
+    def get_cached_instances():
+        return tuple(OrtBackend.__instance_cache)
+
+
+@compatibility(is_backward_compatible=False)
+def make_torch_compile_backend():
+    def compiler_fn(
```

Comment:
nit: outer function `make_torch_compile_backend` can be dropped, and `compiler_fn` can be renamed and passed to `register_backend` directly.

Reply:
Oh, nice. I suppose that can happen now. `make_torch_compile_backend` used to do more 😄